modules/sasl.cpp

```diff
@@ -338,10 +463,256 @@ class CSASLMod : public CModule {
+        pDigestName = "SHA256";
+    } else if (sMechanism.Equals("SCRAM-SHA-512")) {
+        pDigestName = "SHA512";
+    }
+
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
+    OpenSSL_add_all_algorithms();
+#endif
+    pDigest = EVP_get_digestbyname(pDigestName.c_str());
+
+    if (pDigest == nullptr) {
+        throw m_pModule->t_f("Unknown message digest: {1}")(pDigestName);
+    }
+
+    m_pDigest = pDigest;
+    m_uDigestSize = EVP_MD_size(pDigest);
+}
+
+Scram::~Scram() {
+    if (m_pSaltedPassword != nullptr) {
+        free(m_pSaltedPassword);
+    }
+}
+
+int Scram::CreateSHA(const unsigned char* pInput, size_t uInputLen,
+                     unsigned char* pOutput, unsigned int* pOutputLen) {
+    EVP_MD_CTX* pMdCtx = EVP_MD_CTX_new();
+
+    if (!EVP_DigestInit_ex(pMdCtx, m_pDigest, NULL)) {
+        m_pModule->PutModule("Message digest initialization failed");
+        EVP_MD_CTX_free(pMdCtx);
+        return SCRAM_ERROR;
+    }
+
+    if (!EVP_DigestUpdate(pMdCtx, pInput, uInputLen)) {
+        m_pModule->PutModule("Message digest update failed");
+        EVP_MD_CTX_free(pMdCtx);
+        return SCRAM_ERROR;
+    }
+
+    if (!EVP_DigestFinal_ex(pMdCtx, pOutput, pOutputLen)) {
+        m_pModule->PutModule("Message digest finalization failed");
+        EVP_MD_CTX_free(pMdCtx);
+        return SCRAM_ERROR;
+    }
+
+    EVP_MD_CTX_free(pMdCtx);
+    return SCRAM_IN_PROGRESS;
+}
+
+Scram::ScramStatus Scram::ProcessClientFirst() {
+    CString sOutput;
+    m_sClientNonceB64 = CString::RandomString(NONCE_LENGTH);
+    m_sClientNonceB64.Base64Encode();
+    sOutput =
+        "n,,n=" + m_pModule->GetNV("username") + ",r=" + m_sClientNonceB64;
+    m_sClientFirstMessageBare = sOutput.substr(3);
+    sOutput.Base64Encode();
+    m_pModule->SendAuthentication(sOutput);
+    m_uiStep++;
+    return SCRAM_IN_PROGRESS;
+}
+
+Scram::ScramStatus Scram::ProcessServerFirst(CString& sInput) {
+    CString sClientFinalMessageWithoutProof, sServerNonceB64, sSalt,
+        sClientProofB64;
+    VCString vsParams;
+    unsigned char *pClientKey, sStoredKey[EVP_MAX_MD_SIZE], *pClientSignature,
+        *pClientProof;
+    unsigned int uiIndex, uiIterCount = 0, uiClientKeyLen, uiStoredKeyLen;
+    unsigned long ulSaltLen;
+    size_t uClientNonceLen;
+    CString sPassword = m_pModule->GetNV("password");
+    CString sOutput;
+
+    sInput.Split(",", vsParams, false);
+
+    if (vsParams.size() < 3) {
+        m_pModule->PutModule(
+            m_pModule->t_f("Invalid server-first-message: {1}")(sInput));
+        return SCRAM_ERROR;
+    }
+
+    for (const CString& sParam : vsParams) {
+        if (!strncmp(sParam.c_str(), "r=", 2)) {
+            sServerNonceB64 = sParam.substr(2);
+        } else if (!strncmp(sParam.c_str(), "s=", 2)) {
+            sSalt = sParam.substr(2);
+        } else if (!strncmp(sParam.c_str(), "i=", 2)) {
+            uiIterCount = strtoul(sParam.substr(2).c_str(), NULL, 10);
+        }
+    }
+
```

Comment:
This does not properly abort on the presence of `m=`.

Reply:
It also does not verify that the server is not repeating parameters, and it does not abort on the presence of a server error (`e=`).